[MSE] imported/w3c/web-platform-tests/media-source/mediasource-config-change-mp4-av-audio-bitrate.html is a constant timeout.

jyavenard opened the pull request:
#### 9a97db3ec439cfd792a342618a43ce274654a514
<pre>
[MSE] imported/w3c/web-platform-tests/media-source/mediasource-config-change-mp4-av-audio-bitrate.html is a constant timeout.
<a href="https://bugs.webkit.org/show_bug.cgi?id=302695">https://bugs.webkit.org/show_bug.cgi?id=302695</a>
<a href="https://rdar.apple.com/164943486">rdar://164943486</a>

Reviewed by Youenn Fablet.

Many tests has been marked as failed during a 2020 WPT re-sync, mostly due to
a failure to update the expected files.
We update all tests as required and remove the incorrect TestExpectations.
This increase our test coverage of MSE by over 50%

Fly-By: We move the responsability from notifying the MediaPlayerPrivateClient
that all data has been received following a call to MediaSource.endOfStream()
from the MediaSourcePrivate to the MediaPlayerPrivate and use a more clear
nomenclature to indicate what is happening.

Covered by existing tests.
* LayoutTests/imported/w3c/web-platform-tests/media-source/dedicated-worker/mediasource-worker-play-terminate-worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-append-buffer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-endofstream-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-errors-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-getvideoplaybackquality-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-removesourcebuffer-expected.txt:
* LayoutTests/media/media-source/media-source-rendering-update-count-expected.txt:
* LayoutTests/media/media-source/media-source-rendering-update-count.html:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/media-source/SourceBuffer-abort-expected.txt: Removed.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/media-source/SourceBuffer-abort-readyState-expected.txt: Removed.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/media-source/SourceBuffer-abort-removed-expected.txt: Removed.
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::streamEndedWithError):
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::mediaSourceHasRetrievedAllData):
* Source/WebCore/platform/graphics/MediaSourcePrivate.cpp:
(WebCore::MediaSourcePrivate::markEndOfStream):
* Source/WebCore/platform/graphics/MediaSourcePrivate.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h: The HTMLMediaElement shouldn&apos;t emit `stalled` or `progress` event
as when MSE is in use, the user agent never fetches data on behalf of the player.
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::mediaSourceHasRetrievedAllData):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.mm:
(WebCore::MediaSourcePrivateAVFObjC::markEndOfStream): Deleted.
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamer::mediaSourceHasRetrievedAllData):
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h:
* Source/WebCore/platform/graphics/gstreamer/mse/MediaSourcePrivateGStreamer.cpp:
(WebCore::MediaSourcePrivateGStreamer::markEndOfStream):
* Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.cpp:
(WebCore::MockMediaPlayerMediaSource::mediaSourceHasRetrievedAllData):
* Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.h:
* Source/WebCore/platform/mock/mediasource/MockMediaSourcePrivate.cpp:
(WebCore::MockMediaSourcePrivate::markEndOfStream): Deleted.
* Source/WebCore/platform/mock/mediasource/MockMediaSourcePrivate.h:
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.cpp:
(WebKit::MediaSourcePrivateRemote::markEndOfStream):

Canonical link: <a href="https://commits.webkit.org/303737@main">https://commits.webkit.org/303737@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e081fa0979026e95746ad48cc3b6df244de773ce

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133472 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5973 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44616 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141028 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/85520 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7aebd005-f766-4771-83f0-943a99d5b0de) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135342 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6488 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5838 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102101 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/85520 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136419 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/4594 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119648 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82898 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/db178be6-2847-4b33-a8d6-741a61a4ed8d) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4472 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2070 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113590 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37757 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143675 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5643 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38344 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110473 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5725 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4838 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110655 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28047 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4332 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115906 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/59393 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5698 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34223 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5544 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69150 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5787 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5654 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -787,6 +787,10 @@ void MediaSource::streamEndedWithError(std::optional<EndOfStreamError> error)
+
+        // This will fire the `progress` event
+        // Once the entire media resource has been fetched (but potentially before any of it has been decoded)
+        // Fire an event named progress at the media element.
```
Comment in line 791 is somehow redundant with 793

youennf commented on Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
```diff
@@ -179,6 +179,11 @@ class MediaPlayerPrivateInterface : public AbstractRefCounted {
     virtual MediaPlayer::ReadyState readyState() const = 0;
 
+#if ENABLE(MEDIA_SOURCE)
+    // Notify the media element that it now has all of the media data.
```
```suggestion
```

youennf commented on LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-errors-expected.txt:
```diff
@@ -1,8 +1,9 @@
 
-PASS Appending media segment before the first initialization segment.
+FAIL Appending media segment before the first initialization segment. assert_equals: Event types match. expected "error" but got "updateend"
```
It looks like a regression. My understanding is that we are rebasing old expectations that were no longer in sync. Is that correct?

jyavenard replied:
yep.. need to investigate.

jyavenard replied:
the test was still skipped, which is why I didn't notice the error.

jyavenard replied:
We fail the test because the AVStreamDataParser can construct an init segment from the media segment being appended, so this particular content tested doesn't allow for what is being tested.

jyavenard replied:
So the AVStreamDataParser will do a binary scan and detect the AAC raw stream in there as it's the first track of the media segment added; it construct an init segment containing a single audio track rather than throwing an error. I will continue to skip this test for now and leave the expectation as-is

jyavenard committed:
[MSE] imported/w3c/web-platform-tests/media-source/mediasource-config-change-mp4-av-audio-bitrate.html is a constant timeout.
https://bugs.webkit.org/show_bug.cgi?id=302695
rdar://164943486

Reviewed by Youenn Fablet.

Many tests has been marked as failed during a 2020 WPT re-sync, mostly due to
a failure to update the expected files.
We update all tests as required and remove the incorrect TestExpectations.
This increase our test coverage of MSE by over 50%

Fly-By: We move the responsability from notifying the MediaPlayerPrivateClient
that all data has been received following a call to MediaSource.endOfStream()
from the MediaSourcePrivate to the MediaPlayerPrivate and use a more clear
nomenclature to indicate what is happening.

Covered by existing tests.
* LayoutTests/imported/w3c/web-platform-tests/media-source/dedicated-worker/mediasource-worker-play-terminate-worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-append-buffer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-endofstream-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-errors-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-getvideoplaybackquality-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-removesourcebuffer-expected.txt:
* LayoutTests/media/media-source/media-source-rendering-update-count-expected.txt:
* LayoutTests/media/media-source/media-source-rendering-update-count.html:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/media-source/SourceBuffer-abort-expected.txt: Removed.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/media-source/SourceBuffer-abort-readyState-expected.txt: Removed.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/media-source/SourceBuffer-abort-removed-expected.txt: Removed.
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::streamEndedWithError):
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::mediaSourceHasRetrievedAllData):
* Source/WebCore/platform/graphics/MediaSourcePrivate.cpp:
(WebCore::MediaSourcePrivate::markEndOfStream):
* Source/WebCore/platform/graphics/MediaSourcePrivate.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h: The HTMLMediaElement shouldn't emit `stalled` or `progress` event
as when MSE is in use, the user agent never fetches data on behalf of the player.
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::mediaSourceHasRetrievedAllData):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.mm:
(WebCore::MediaSourcePrivateAVFObjC::markEndOfStream): Deleted.
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamer::mediaSourceHasRetrievedAllData):
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h:
* Source/WebCore/platform/graphics/gstreamer/mse/MediaSourcePrivateGStreamer.cpp:
(WebCore::MediaSourcePrivateGStreamer::markEndOfStream):
* Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.cpp:
(WebCore::MockMediaPlayerMediaSource::mediaSourceHasRetrievedAllData):
* Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.h:
* Source/WebCore/platform/mock/mediasource/MockMediaSourcePrivate.cpp:
(WebCore::MockMediaSourcePrivate::markEndOfStream): Deleted.
* Source/WebCore/platform/mock/mediasource/MockMediaSourcePrivate.h:
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.cpp:
(WebKit::MediaSourcePrivateRemote::markEndOfStream):

Canonical link: https://commits.webkit.org/303737@main
